Add randomness contract support to AuthorityRound.

afck opened the pull request:
This allows `AuthorityRound` validators to participate in a [RANDAO](https://github.com/randao/randao)-like RNG smart contract. We implemented it as part of [POSDAO](https://forum.poa.network/t/posdao-white-paper/2208), but it would work as a standalone feature to provide on-chain randomness, too.

See the description below: https://github.com/paritytech/parity-ethereum/pull/10946#issuecomment-547857765

Original implementation by @mbr: https://github.com/poanetwork/parity-ethereum/pull/52

I'm not really happy with the fact that we can only call e.g. `isCommitPhase` on the "latest" block, i.e. the _parent_ of the one we're currently creating, and we're including our transactions in. This means authors of randomness contracts need to check in `commitHash` whether `_isCommitPhase(block.number - 1)`.
For a different feature, we actually added a `Client::call_contract_before` method that allows you to make calls at the current pending block: https://github.com/poanetwork/parity-ethereum/pull/140/files#diff-40c7c65265e511d07158f62712f69573R1460
I.e. the call is made at the initial state of the pending block, but with the correct new block number. Do you think it would make sense to add that to `Client` and use it here instead?

Another issue is that without `--force-sealing`, we're not producing new blocks, even if it would be our turn to send a value to the randomness contract. I'm not sure where the right place in the code is to put a warning in that case; or where I'd even have access to both `force_sealing` and `randomness_contract_address`?

Is there any place where I should add documentation for the contract API?

Please let me know your thoughts!

afck commented:
I can't reproduce that failure locally:
```
$ cargo test should_check_status_of_request_when_its_resolved
    Finished dev [unoptimized + debuginfo] target(s) in 0.21s
     Running /home/andreas/git/parity-ethereum/target/debug/deps/parity_rpc-36b54cdf58daa398

running 1 test
test v1::tests::mocked::signing::should_check_status_of_request_when_its_resolved ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 297 filtered out
```

afck commented:
Hi! I'm still interested in merging this, and happy address any concerns about the implementation.

varasev commented on ethcore/res/contracts/authority_round_random.json:
```diff
@@ -0,0 +1,158 @@
+[{
+		"constant": true,
+		"inputs": [],
+		"name": "currentRandom",
```
@afck Let's remove `currentRandom` from the Random contract ABI since we don't use that (and we replaced that with [`uint256 public currentSeed`](https://github.com/poanetwork/posdao-contracts/blob/claim-reward/contracts/RandomAuRa.sol#L28)).

afck replied:
Done.

varasev commented on ethcore/engines/authority-round/src/lib.rs:
```diff
@@ -1430,6 +1452,46 @@ impl Engine for AuthorityRound {
 		block_reward::apply_block_rewards(&rewards, block, &self.machine)
 	}
 
+	/// Make calls to the randomness and validator set contracts.
```
We don't call the `ValidatorSet` contract here, so this comment should be changed.

afck replied:
Fixed.

varasev commented:
Hi everyone,

I'd like to clarify the logic of implemented RANDAO-like mechanism to help understanding the suggested code.

There are so-called `collection rounds` which have a fixed length in blocks. For example, POSDAO's collection round length will be `38` blocks.

Each collection round is split into two equal phases: `commit phase` and `reveal phase`. The length of each phase is equal to `19` blocks in the above case (because `38 / 2 = 19`).

When a collection round finishes, the next collection round starts, and so on. For example:

```
Block number   |   Phase
...
101                Commit phase     <-- collection round #1 started here
102                Commit phase
103                Commit phase
...
117                Commit phase
118                Commit phase
119                Commit phase
120                Reveal phase
121                Reveal phase
122                Reveal phase
...
136                Reveal phase
137                Reveal phase
138                Reveal phase     <-- collection round #1 finished here
139                Commit phase     <-- collection round #2 started here
140                Commit phase
141                Commit phase
...
```

There is a smart contract (let's name it `RandomAuRa`) which collects the random numbers (generated by validators) on each `collection round`.

During the `commit phase` each validator in the validator set generates a `secret` and calls a [`RandomAuRa.commitHash`](https://github.com/poanetwork/posdao-contracts/blob/29959a76687b5bd79ac1a3c6c1673b4c9e011396/contracts/RandomAuRa.sol#L62) function when their turn (i.e. when the validator creates a block):

```solidity
function commitHash(bytes32 _secretHash, bytes calldata _cipher) external;
```

This function accepts the hash of the `secret` and its `cipher`. The `cipher` is the `secret` encrypted with a validator's key. This is needed for the `reveal phase` (see below).

For example, if there are three validators, they will call `commitHash` on the following order (for the sample case above):

```
Block number   |   Phase   |       What happens
...
101             Commit phase    Validator1 calls `commitHash`
102             Commit phase    Validator2 calls `commitHash`
103             Commit phase    Validator3 calls `commitHash`
104             Commit phase    Nothing happens
105             Commit phase    Nothing happens
106             Commit phase    Nothing happens
...
117             Commit phase    Nothing happens
118             Commit phase    Nothing happens
119             Commit phase    Nothing happens
120             Reveal phase    Validator1 calls `revealSecret`
121             Reveal phase    Validator2 calls `revealSecret`
122             Reveal phase    Validator3 calls `revealSecret`
123             Reveal phase    Nothing happens
125             Reveal phase    Nothing happens
126             Reveal phase    Nothing happens
...
136             Reveal phase    Nothing happens
137             Reveal phase    Nothing happens
138             Reveal phase    Nothing happens
139             Commit phase    Validator1 calls `commitHash`
140             Commit phase    Validator2 calls `commitHash`
141             Commit phase    Validator3 calls `commitHash`
...
```

So, when the `commit phase` finishes, the `reveal phase` starts, and the validators call the [`RandomAuRa.revealSecret`](https://github.com/poanetwork/posdao-contracts/blob/29959a76687b5bd79ac1a3c6c1673b4c9e011396/contracts/RandomAuRa.sol#L79) function to reveal their committed secrets:

```solidity
function revealSecret(uint256 _secret) external;
```

Before calling this, a validator reads the cipher of the `secret` from the contract with [`RandomAuRa.getCipher`](https://github.com/poanetwork/posdao-contracts/blob/29959a76687b5bd79ac1a3c6c1673b4c9e011396/contracts/RandomAuRa.sol#L196) public getter:

```solidity
function getCipher(uint256 _collectRound, address _miningAddress) public view returns(bytes memory);
```

And then the validator decrypts the cipher with their key, retrieves the `secret`, and calls the `revealSecret`.

The code in this PR also uses the following public getters of the `RandomAuRa` contract to determine whether a validator should call `commitHash` or `revealSecret` in the current block (or shouldn't call them in the current collection round anymore):

```solidity
/// @dev Returns the serial number of the current collection round.
/// Needed when using `getCommit`, `isCommitted`, or `sentReveal` getters (see below).
function currentCollectRound() public view returns(uint256);

/// @dev Returns the Keccak-256 hash of the validator's secret for the specified collection round and the specified
/// validator stored by the validator through the `commitHash` function.
/// @param _collectRound The serial number of the collection round for which the hash should be retrieved.
/// Should be read with `currentCollectRound()` getter.
/// @param _miningAddress The mining address of validator (engine_signer).
function getCommit(uint256 _collectRound, address _miningAddress) public view returns(bytes32);

/// @dev Returns a boolean flag indicating whether the specified validator has committed their secret's hash for the
/// specified collection round.
/// @param _collectRound The serial number of the collection round for which the checkup should be done.
/// Should be read with `currentCollectRound()` getter.
/// @param _miningAddress The mining address of the validator (engine_signer).
function isCommitted(uint256 _collectRound, address _miningAddress) public view returns(bool);

/// @dev Returns a boolean flag of whether the specified validator has revealed their secret for the
/// specified collection round.
/// @param _collectRound The serial number of the collection round for which the checkup should be done.
/// Should be read with `currentCollectRound()` getter.
/// @param _miningAddress The mining address of the validator (engine_signer).
function sentReveal(uint256 _collectRound, address _miningAddress) public view returns(bool);

/// @dev Returns a boolean flag indicating whether the current phase of the current collection round
/// is a `commit phase`. Used by the validator's node to determine if it should commit the hash of
/// the secret during the current collection round.
function isCommitPhase() public view returns(bool);

/// @dev Returns a boolean flag indicating whether the current phase of the current collection round
/// is a `reveal phase`. Used by the validator's node to determine if it should reveal the secret during
/// the current collection round.
function isRevealPhase() public view returns(bool);
```

afck commented:
Yes, I think we need those changes for two things:
* making zero-gas-price transactions, and
* making transactions without pushing them to the queue, so that we can directly put them into a block we are creating.

I'm happy to do these differently, if there are better ways to do it.

dvdplm commented on json/src/spec/authority_round.rs:
```diff
@@ -105,6 +107,17 @@ pub struct AuthorityRound {
+#[derive(Debug, PartialEq, Deserialize)]
+#[serde(deny_unknown_fields)]
+#[serde(untagged)]
+pub enum TransitionMap<T> {
```
As this is a new feature I wonder if we really need to be this fancy. Can't we mandate a map here and avoid this?

afck replied:
The [`posdao-contracts` chainspec template](https://github.com/poanetwork/posdao-contracts/blob/6538a3c2e5bbb09f41cf1fb7c28b2405ed35ee8c/templates/spec.json#L19) is an example for how we are currently using this.
If we allow not using a map, it looks just like e.g. the block reward contract.

varasev replied:
Yes, we suppose that the address of randomness contract can be changed in spec, so we decided to have that as a map, e.g.:

```
"randomnessContractAddress": {
  "0": "0xaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
  "100": "0xbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb"
}
```

But we also suggest to have a simpler form of first address definition, e.g. when we define it as

```
"randomnessContractAddress": {
  "0": "0xaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
}
```

we could write that as

```
"randomnessContractAddress": "0xaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
```

dvdplm replied:
I agree that it'd be nice to let knowledgeable users to have a short-form for the single-item use case. However I also think that it's worth more to have less rust code to maintain: configuration mistakes are non-catastrophic (the node won't start; bugs in rust might be more subtle) and probably easy to mitigate with docs. So yeah, unless you disagree strongly I think it's good to mandate a map here.

varasev replied:
This is addressed in https://github.com/poanetwork/parity-ethereum/commit/4799aeb3d11bac2910ff04e5714631a541d92aec.

dvdplm commented on ethcore/src/miner/miner.rs:
```diff
@@ -503,12 +526,11 @@ impl Miner {
+		debug!(target: "miner", "Attempting to push {} transactions.", engine_pending.len() + queue_pending.len());
 
-		for tx in pending {
+		for transaction in engine_pending.into_iter().chain(queue_pending.into_iter().map(|tx| tx.signed().clone())) {
```
Here there's a potential problem I think: if the engine's txs can't **all** be included in the block we should…  do what? wait for the next block? 
Either way: as it stands, unless all the engine's txs are included there's a problem right? Maybe add them separately (instead of `.chain()`-ing them with the normal txs)?

afck replied:
I'm still hesitating:
If we merge #11245, this will be used for queued malice reports. If those don't all fit into a block, that's not catastrophic at all: They'd just be retried. But if we return an error here, we'd be unable to ever produce a block again in that scenario.

We could instead allow `on_prepare_block` to mark its returned transactions as critical. Then block creation would only fail if a _critical_ transaction failed to be included.

dvdplm commented on ethcore/src/miner/miner.rs:
```diff
@@ -432,36 +442,49 @@ impl Miner {
 			}
 
-			(open_block, last_work_hash)
+			last_work_hash
```
This looks odd. Needs some refactoring for readability.

dvdplm commented on ethcore/src/test_helpers/test_client.rs:
```diff
@@ -912,18 +912,37 @@ impl BlockChainClient for TestBlockChainClient {
 	}
 
-	fn transact_contract(&self, address: Address, data: Bytes) -> Result<(), transaction::Error> {
+	fn create_transaction(
```
Document why this passes in nonce and gas price from the outside.

varasev commented on ethcore/engines/authority-round/src/randomness.rs:
```diff
@@ -0,0 +1,245 @@
+	/// Calls various constant contract functions to determine the precise state that needs to be
+	/// handled (that is, the phase and whether or not the current validator still needs to send
+	/// commitments or reveal secrets).
+	pub fn load(
```
`current_collect_round()` is needed for `is_committed()` and `sent_reveal()` since they accept `_collectRound` parameter.

We use `is_committed()` and `sent_reveal()` to prevent duplicated calls to `commit_hash()` and `reveal_secret()` respectively.

To simplify that a bit we could utilize `is_commit_phase()` instead of `is_reveal_phase()` as `is_commit_phase() == !is_reveal_phase()`.

varasev replied:
@afck Let's use `is_commit_phase()` instead of `is_reveal_phase()`: `let is_reveal_phase = !is_commit_phase` in https://github.com/poanetwork/parity-ethereum/blob/727fce39a1b8c922db27533b9d38bc879f707a32/ethcore/engines/authority-round/src/randomness.rs#L146-L151

varasev replied:
`is_reveal_phase` was removed from the [`load` function](https://github.com/poanetwork/parity-ethereum/blob/4799aeb3d11bac2910ff04e5714631a541d92aec/ethcore/engines/authority-round/src/randomness.rs#L147).

varasev commented on ethcore/engines/authority-round/src/lib.rs:
```diff
@@ -1430,6 +1452,46 @@ impl Engine for AuthorityRound {
+		let full_client = client.as_full_client()
+			.ok_or(EngineError::FailedSystemCall("Failed to upgrade to BlockchainClient.".to_string()))?;
+
+		// Our current account nonce. The transactions must have consecutive nonces, starting with this one.
```
Seems this (obsolete) comment relates to implementation in our fork where we also performed other transactions (like calling `emitInitiateChange` function).

dvdplm replied:
I see, so maybe somewhat related to https://github.com/paritytech/parity-ethereum/pull/11245?

varasev replied:
Yes, here: https://github.com/paritytech/parity-ethereum/pull/11245/files#diff-79ef4a3dce58a9d303d49c5aee781e17R1484

dvdplm replied:
If you don't mind, let's try to keep this PR focussed on the RANDAO changes and take the next one once this one lands? (if possible ofc)

varasev replied:
Yes, of course. I just clarified why the code assumes having multiple transactions. Seems that needs to be fixed here. Hope @afck get well soon and fix that. He told he's a bit ill at the moment

varasev replied:
@afck So, I guess we should get rid of the `transactions` vector [here](https://github.com/poanetwork/parity-ethereum/blob/727fce39a1b8c922db27533b9d38bc879f707a32/ethcore/engines/authority-round/src/lib.rs#L1087-L1107) since there is only one transaction in this PR? Or add a comment that the vector can be needed for other transactions in the future?

afck replied:
Done in https://github.com/paritytech/parity-ethereum/pull/10946/commits/0d7d0c8ecd13f658b9299fde19e9e15ebf1e788f.

varasev commented on ethcore/engines/authority-round/src/randomness.rs:
```diff
@@ -0,0 +1,245 @@
+				Ok(Some(data))
+			}
+			RandomnessPhase::Reveal { round, our_address } => {
+				// Load the hash and encrypted secret that we stored in the commit phase.
```
Ok, let's use `RandomAuRa.getCommitAndCipher()` instead: https://github.com/poanetwork/posdao-contracts/commit/696fdd193319cb1afa21e7b96d3b82dbaec566c8

varasev replied:
Done in https://github.com/poanetwork/parity-ethereum/commit/624ce47fdb8d3621701afa23614ec2b39aba62d1.

varasev commented on ethcore/engines/authority-round/src/randomness.rs:
```diff
@@ -0,0 +1,245 @@
+//! This module contains the support code for the on-chain randomness generation used by AuRa. Its
+//! core is the finite state machine `RandomnessPhase`, which can be loaded from the blockchain
+//! state, then asked to perform potentially necessary transaction afterwards using the `advance()`
+//! method.
```
The current contract's code could be got from [my above comment](https://github.com/paritytech/parity-ethereum/pull/10946#issuecomment-548404697).

varasev replied:
It seems the comments were expanded: https://github.com/poanetwork/parity-ethereum/blob/727fce39a1b8c922db27533b9d38bc879f707a32/ethcore/engines/authority-round/src/randomness.rs#L19-L58

varasev commented on ethcore/engines/authority-round/src/randomness.rs:
```diff
@@ -0,0 +1,245 @@
+/// Secret type expected by the contract.
+// Note: Conversion from `U256` back into `[u8; 32]` is cumbersome (missing implementations), for
+//       this reason we store the raw buffers.
+pub type Secret = [u8; 32];
```
> I guess then we should change the method names in the contract as well, e.g. `reveal_secret`, etc.?

@afck, let's leave the names in the contracts as they are. I guess, David meant that `Secret` is already used in Parity's code, but I believe we should keep `RANDAO` terminology in the contracts.

> Using `H256` instead of a byte array is also something we'd need to change in the contract

Sorry, do you mean `bytes _cipher` in the `RandomAuRa.commitHash` function?

The [secret](https://github.com/poanetwork/posdao-contracts/blob/696fdd193319cb1afa21e7b96d3b82dbaec566c8/contracts/RandomAuRa.sol#L79) and [`currentSeed`](https://github.com/poanetwork/posdao-contracts/blob/696fdd193319cb1afa21e7b96d3b82dbaec566c8/contracts/RandomAuRa.sol#L28) in `RandomAuRa` contract are already `uint256`.

varasev replied:
Or you meant that the `secret` in the contract should be `bytes32`? If so, could you convert that to `uint256` on Parity's side to keep the contract compatible with our `aura-pos` fork?

dvdplm replied:
@varasev It is true that `Secret` is already a type in Parity, but I guess my objection was also around the role it plays in the RANDAO process (as I understood it, I might be completely wrong!): it is a secret piece of data for sure, but its main property is that it is the piece of randomness that this node contributes to the final random number. In other words: the name doesn't reflect its main purpose.

varasev replied:
@dvdplm Ok. Would you suggest renaming `revealSecret` to `revealNumber` in the contract?

afck replied:
Right, sorry! I was confused because we're calling `revealSecret` with an `[u8; 32]` argument in the code, but that just works because it implements `Into<U256>`. The contract ABI is already fine then. :+1: 
(Apart from possibly the names.)

afck replied:
Actually… `H256` would be better than `U256`. It's not interpreted as a number, so it has an [`as_fixed_bytes`](https://docs.rs/ethereum-types/0.7.0/ethereum_types/struct.H256.html#method.as_fixed_bytes) method and no endianness.

varasev replied:
@dvdplm I renamed `RandomAuRa.revealSecret` function to `revealNumber` in the code, in the contracts, and in my above comments:

https://github.com/paritytech/parity-ethereum/pull/10946/commits/eaa21315a5d43bba4661cb64ecc765a5c6ad7845
https://github.com/poanetwork/posdao-contracts/commit/ec3901160757ba295efd273a5cbd543037ceb1ce
https://github.com/paritytech/parity-ethereum/pull/10946#issuecomment-547857765
https://github.com/paritytech/parity-ethereum/pull/10946#issuecomment-548404697

varasev commented on ethcore/engines/authority-round/src/randomness.rs:
```diff
@@ -0,0 +1,245 @@
+
+				trace!(target: "engine", "Randomness contract: revealing secret for {}.", secret_hash);
+				// We are now sure that we have the correct secret and can reveal it.
+				let (data, _decoder) = aura_random::functions::reveal_secret::call(secret);
```
Yes, the data is returned [here](https://github.com/poanetwork/parity-ethereum/blob/bbe6c24fb67e99e900f206962fb153f6ce6ee355/ethcore/engines/authority-round/src/lib.rs#L1103-L1106) and then the prepared transactions are included in a block [here](https://github.com/poanetwork/parity-ethereum/blob/bbe6c24fb67e99e900f206962fb153f6ce6ee355/ethcore/src/miner/miner.rs#L472).

afck commented on ethcore/engines/authority-round/src/randomness.rs:
```diff
@@ -0,0 +1,245 @@
+				let committed_hash: Hash = contract
+					.call_const(aura_random::functions::get_commit::call(round, our_address))
+					.map_err(PhaseError::LoadFailed)?;
+				if !committed_hash.is_zero() {
```
Yes, I'm happy to remove it.
This code was written very defensively because it's supposed to work with randomness contract implementations that are not part of this repository. It would be a bug, but a contract could, of course, trigger this error.

varasev replied:
The contract takes into account this (see the implementation of `commitHash` function in https://github.com/paritytech/parity-ethereum/pull/10946#issuecomment-548404697).

varasev replied:
That check was removed.

afck commented on ethcore/engines/authority-round/src/randomness.rs:
```diff
@@ -0,0 +1,245 @@
+				let secret = if secret_vec.len() == 32 {
+					let mut buf = [0u8; 32];
+					buf.copy_from_slice(&secret_vec);
+					buf
```
I think `from_slice` also copies the data, though. Would be great if `H256` had a `from_vec`.

afck commented:
> Tabs and spaces are mixed here point_up

Where? I can't find it.

afck commented on ethcore/engines/authority-round/src/randomness.rs:
```diff
@@ -0,0 +1,245 @@
+	LoadFailed(CallError),
+	/// Failed to load the stored secret.
+	#[display(fmt = "Failed to load secret from the randomness contract")]
+	StaleSecret,
```
I renamed it to `BadRandNumber`.

afck commented on ethcore/engines/authority-round/src/randomness.rs:
```diff
@@ -0,0 +1,245 @@
+	Crypto(CryptoError),
+	/// Failed to decrypt stored secret.
+	#[display(fmt = "Failed to decrypt stored randomness secret: {}", _0)]
+	Decrypt(CryptoError),
```
Done.

afck commented on ethcore/src/miner/miner.rs:
```diff
@@ -419,9 +427,11 @@ impl Miner {
 
 		// Open block
-		let (mut open_block, original_work_hash) = {
+		let original_work_hash = {
```
I went with the latter: Looks like `original_work_hash` was always set to `last_work_hash` anyway.

ordian commented on ethcore/engines/authority-round/src/lib.rs:
```diff
@@ -1045,6 +1068,46 @@ impl AuthorityRound {
+		// Creates and signs a transaction with the given contract call.
+		let mut make_transaction = |to: Address, data: Bytes| -> Result<SignedTransaction, Error> {
+			let nonce = Some(tx_nonce);
+			tx_nonce += U256::one(); // Increment the nonce for the next transaction.
```
is it okay to panic here on overflow? or is it not feasible to get an overflow here?

afck replied:
Is there a way to artificially increase account nonces somehow?
Otherwise I'd say it's not feasible: You'd have to execute 2^256 transactions to get there.
Not sure whether the yellow paper even defines what should happen in that case? At a glance, I couldn't find it.

varasev commented on ethcore/engines/authority-round/src/lib.rs:
```diff
@@ -1430,6 +1452,46 @@ impl Engine for AuthorityRound {
 		block_reward::apply_block_rewards(&rewards, block, &self.machine)
 	}
 
+	/// Make calls to the randomness contract.
```
It seems @afck moved the function to `run_randomness_phase`: https://github.com/poanetwork/parity-ethereum/blob/727fce39a1b8c922db27533b9d38bc879f707a32/ethcore/engines/authority-round/src/lib.rs#L1073

dvdplm commented on ethcore/engines/authority-round/src/lib.rs:
```diff
@@ -1430,6 +1452,46 @@ impl Engine for AuthorityRound {
 	}
 
+	/// Make calls to the randomness contract.
+	fn on_prepare_block(&self, block: &ExecutedBlock) -> Result<Vec<SignedTransaction>, Error> {
```
> Is it still guaranteed at this point that the client's latest nonce is the same, i.e. that the latest block as seen by the client is the parent of this `block`?

I honestly do not know. @tomusdrw maybe?

tomusdrw replied:
Yes, the `nonce` should be the same since we are only calling this function before any transaction is applied. I think we should actually consider renaming this function to `fn generate_engine_transactions(&self, block: &ExecutedBlock) -> Result<Vec<_>>` and document it properly. Having `on_prepare_block` return a list of transactions doesn't seem intuitive for me.

afck replied:
I renamed it.
There is a doc comment on the method in the trait definition.

dvdplm commented on ethcore/engines/authority-round/src/randomness.rs:
```diff
@@ -0,0 +1,248 @@
+	/// Waiting for the next phase.
+	///
+	/// This state indicates either the successful revelation in this round or having missed the
+	/// window to make a commitment.
```
Maybe dumb q: what is the window? A range of blocks? Where do we set this? In the `aura_random` contract? (If yes, I think it's a good idea to add notes for any piece of data that is defined in the contract as opposed to the rust code).

afck replied:
That's up to the contract. In POSDAO it's a range of blocks, yes.
I'll add a comment.

varasev replied:
The comment was added in https://github.com/poanetwork/parity-ethereum/commit/4799aeb3d11bac2910ff04e5714631a541d92aec.

dvdplm commented on ethcore/engines/authority-round/src/randomness.rs:
```diff
@@ -0,0 +1,248 @@
+	///
+	/// Returns the encoded contract call necessary to advance the randomness contract's state.
+	///
+	/// **Warning**: The `advance()` function should be called only once per block state; otherwise
```
Not sure what you mean by "block state"? Can you elaborate?

afck replied:
I think the intent of this warning was to make sure we don't call `advance()` twice in the same block. I'll reword it.

dvdplm commented on ethcore/engines/authority-round/src/util.rs:
```diff
@@ -0,0 +1,78 @@
+
+/// Contract call failed error.
+#[derive(Debug)]
+pub enum CallError {
```
@afck are you planning on addressing this or do you disagree?

afck replied:
Yes, I'll give it a try. :+1:

afck replied:
I do agree in general, but I'm unsure what to do exactly:

Replacing `call-contract`'s `String` with a proper error type seems orthogonal to this PR: In the `authority-round/util` module, we'd still have two additional error cases, namely that the client is not a full client, and that decoding the call's output fails. We'd only replace the `String` in the `CallFailed` variant.

So if all we want to change is `call-contract`'s error type, I'd do that in a separate PR.

But I'm unsure about the whole `util` module in general: We were struggling a lot trying to find out what's "the right way" to make contract calls in Parity. Maybe `util`'s functionality should be part of the `Client` implementation? Or part of the `call-contract` crate?

dvdplm replied:
> So if all we want to change is `call-contract`'s error type, I'd do that in a separate PR.

Fair enough. And I hear you on the difficulties finding the "proper" place for things. My gut feeling is we'd need a pretty deep redesign but yeah, I don't have a good answer or rule of thumb for the time being. If at all possible I'd try to keep code out of `Client`. Calling contracts might be general enough to warrant the error type to live in `common-types` too, that's what we've done for other generally useful error types.

dvdplm replied:
Thank you for looking into it though!

dvdplm commented on ethcore/engines/authority-round/src/util.rs:
```diff
@@ -0,0 +1,94 @@
+
+	/// Perform a function call to an ethereum machine that doesn't create a transaction or change the state.
+	///
+	/// Runs a constant function call on `client`. The `call` value can be serialized by calling any
```
I think I've asked this somewhere else already but I do not remember the answer, apologies. What is meant by a "constant function" in this context? Is it "constant" because it does not create txs and thus does not alter state? If yes, I suggest being explicit: "Runs a non-state altering function call…" (or something to that effect).

afck replied:
Yes, exactly. I'll expand the comment.

varasev replied:
The comment was expanded in https://github.com/poanetwork/parity-ethereum/commit/4799aeb3d11bac2910ff04e5714631a541d92aec#diff-a4faad7e2fe2f519b612d48dfa0f8607.

afck commented:
I addressed most comments, except the error-related ones (also https://github.com/paritytech/parity-ethereum/pull/10946#discussion_r346882321 — see my reply; happy to work on this once we make a decision).

afck commented on ethcore/engines/authority-round/src/util.rs:
```diff
@@ -0,0 +1,78 @@
+impl<'a> BoundContract<'a> {
+	/// Create a new `BoundContract`.
+	#[inline]
+	pub fn bind(client: &dyn EngineClient, block_id: BlockId, contract_addr: Address) -> BoundContract {
```
Done.

varasev commented on ethcore/src/miner/miner.rs:
```diff
@@ -419,9 +427,11 @@ impl Miner {
 	{
 		trace_time!("prepare_block");
 		let chain_info = chain.chain_info();
+		let engine_pending;
```
The comment was added in https://github.com/paritytech/parity-ethereum/commit/0d7d0c8ecd13f658b9299fde19e9e15ebf1e788f.

tomusdrw commented on ethcore/types/src/errors/engine_error.rs:
```diff
@@ -35,8 +35,16 @@ pub enum EngineError {
 	/// Validation proof insufficient.
 	InsufficientProof(String),
+	/// Randomness error in load method
+	RandomnessLoadError(String),
```
Seems that the AuRa internal randomness contract is leaking to a generic `EngineError` type, I think we should rather keep these under some more generic variant, especially that we never really handle them explicitly.

afck replied:
Yes, I'm really unsure about how we should organize these errors (also see https://github.com/paritytech/parity-ethereum/pull/10946#discussion_r346882321).

Should we call this `Other(String)` then? Or `Other<Box<dyn Error>>`?

tomusdrw replied:
Yeah, either of those. I think I'm more in favor of simpler `Other(String)`. Ideally would be if `Engine` had an associated type of it's custom error type, but that's way beyond the scope of this PR.

@dvdplm are you fine with `Other(String)` or do you like `Other<Box<dyn Error>>` more?

afck replied:
Actually, there's already a `Custom(String)`.

dvdplm replied:
Let's go with the existing `Custom(String)`. Not ideal, but I agree with Tomek here: improving the error handling here is non-trivial and not in scope for this PR.

afck replied:
Great! I used `Custom` in https://github.com/paritytech/parity-ethereum/pull/10946/commits/ab169e24b8a52539b4edef619517b347d99e4dfb.

varasev commented on ethcore/engines/authority-round/src/randomness.rs:
```diff
@@ -0,0 +1,257 @@
+				let number: RandNumber = rng.gen();
+				let number_hash: Hash = keccak(number.as_bytes());
+				let public = signer.public().ok_or(PhaseError::MissingPublicKey)?;
+				let cipher = ecies::encrypt(&public, &number_hash.0, number.as_bytes())?;
```
We think it would be more reliable to store the cipher on-chain, and, since we call the `commitHash` function anyway, we pass the `cipher` to it at the same time. So storing it on-chain doesn't require making a separate transaction.

afck replied:
Also, if your node crashes somehow and you start another one, it can just continue, and won't fail to reveal the secret.
If we stored that locally, the backup node wouldn't have the required information.

tomusdrw replied:
Would be good to check with some cryptographer if we don't leak some information everytime we encode the same number. I think it's unlikely, but still.

afck replied:
You mean because we publish multiple pairs of plaintext + ciphertext for the same key?
Since [ECIES](https://en.wikipedia.org/wiki/Integrated_Encryption_Scheme) is supposed to be secure even against chosen-plaintext and chosen-ciphertext attacks, I believe that shouldn't be a problem. But I'm not a cryptographer either.

afck commented on ethcore/src/miner/miner.rs:
```diff
@@ -327,6 +327,13 @@ impl Miner {
+	/// Creates new instance of miner with given spec and accounts.
+	///
+	/// NOTE This should be only used for tests.
+	pub fn new_for_tests_force_sealing(spec: &Spec, accounts: Option<HashSet<Address>>, force_sealing: bool) -> Miner {
```
There seem to be two tests relying on it:
```
failures:
    miner::miner::tests::should_not_mine_if_internal_sealing_is_disabled
    miner::miner::tests::should_not_mine_if_no_fetch_work_request
```
We can add the parameter to `new_for_tests`, but that will touch a lot of files.

tomusdrw replied:
Okay, would be best to have some way for them to alter the options, without introducing those additional methods, but I'm fine as-is.

afck commented on ethcore/client-traits/src/lib.rs:
```diff
@@ -388,7 +388,27 @@ pub trait BlockChainClient:
+		&self,
+		action: Action,
+		data: Bytes,
+		gas: Option<U256>,
```
Done…
I didn't separate builder from request, though; let me know if you're okay with that approach.

afck commented on ethcore/engines/authority-round/src/lib.rs:
```diff
@@ -72,13 +74,17 @@ use common_types::{
 	errors::{BlockError, EthcoreError as Error, EngineError},
+	ids::BlockId,
 	snapshot::Snapshotting,
+    transaction::{Action, SignedTransaction},
```
Fixed. (Sorry, somehow sometimes my Vim doesn't realize it's tabs.)

afck commented on ethcore/engines/authority-round/src/lib.rs:
```diff
@@ -1045,6 +1062,41 @@ impl AuthorityRound {
+			EngineError::RequiresClient
+		})?;
+		let full_client = client.as_full_client()
+			.ok_or(EngineError::FailedSystemCall("Failed to upgrade to BlockchainClient.".to_string()))?;
```
Done.

afck commented on ethcore/engines/authority-round/src/util.rs:
```diff
@@ -0,0 +1,95 @@
+
+impl<'a> BoundContract<'a> {
+	/// Create a new `BoundContract`.
+	#[inline]
```
Done.

afck commented on ethcore/src/miner/miner.rs:
```diff
@@ -422,48 +430,61 @@ impl Miner {
-							warn!(target: "miner", "Open new block failed with error {:?}. This is likely an error in chain specificiations or on-chain consensus smart contracts.", err);
-							return None;
-						}
+		let mut sealing = self.sealing.lock();
```
Done.

tomusdrw commented on ethcore/client-traits/src/lib.rs:
```diff
@@ -397,30 +397,69 @@ pub trait BlockChainClient:
 
 	/// Schedule state-altering transaction to be executed on the next pending block.
 	fn transact_contract(&self, address: Address, data: Bytes) -> Result<(), transaction::Error> {
```
I'd remove that method now, it's really easy to use `transact` instead.

afck replied:
I removed it. :+1:

tomusdrw replied:
Perfect, thanks!
